[SPMD] Upstream iter_move_grads_and_optimizers

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98785

This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit to @lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

fegin committed:
[SPMD] Upstream iter_move_grads_and_optimizers

This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863 is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Upstream iter_move_grads_and_optimizers"


This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit for lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

[ghstack-poisoned]

mrshenli commented on test/distributed/_spmd/test_transformation.py:
```diff
@@ -228,6 +231,37 @@ def train_step(model, optim, batch):
+        )
+        for _ in range(num_iters):
+            batch = torch.randn(batch_size, dim).cuda()
+            out = train_step(model, optim, batch)
```
same comment regarding accuracy tests as previous PRs

fegin replied:
Done

fegin committed:
Update on "[SPMD] Upstream iter_move_grads_and_optimizers"


This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit for lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Upstream iter_move_grads_and_optimizers"


This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit for lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Upstream iter_move_grads_and_optimizers"


This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit for lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -890,3 +891,49 @@ def split_fused_optimizer(
+    else:
+        raise ValueError(f"Cannot find {target_comm_node}")
+
+    optim_blocks = get_all_fused_optimizer_blocks(gm, "_fused_adam")
```
this is very specific to adam. Let's add a TODO/FIXME to track

mrshenli commented on torch/distributed/_spmd/graph_utils.py:
```diff
@@ -104,6 +104,20 @@ def get_output(graph: fx.Graph) -> fx.Node:
+    Take a predicate and return all the nodes in the `graph` where the predicate
+    holds.
+    """
+    nodes = graph.nodes
+    if reverse_order:
+        nodes = reversed(nodes)
```
nit: simplification?

```python
nodes = reversed(graph.nodes) if reverse_order else graph.nodes
```

fegin replied:
This is less easier to read after typing hints. Will keep the original form.

mrshenli reviewed (approved):
This is not user-facing yet, stamp to unblock. Please fix Lint before land

fegin commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -890,3 +891,49 @@ def split_fused_optimizer(
+@graph_optimization_pass(run_after=("schedule_comm_wait", "remove_copy_from_optimizer"))
+def iter_move_grads_and_optimizers(
+    gm: IterGraphModule,
+    target_comm_node: str,
+    target_dest_node: str,
```
You are right about the usage of this pass. This enables the ability to move nodes around but the algorithm is not decided yet. Right now the API is intended to be used for manual tuning.

fegin committed:
Update on "[SPMD] Upstream iter_move_grads_and_optimizers"


This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit for lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Upstream iter_move_grads_and_optimizers"


This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit for lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Upstream iter_move_grads_and_optimizers"


This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit for lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Upstream iter_move_grads_and_optimizers"


This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit for lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Upstream iter_move_grads_and_optimizers"


This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit for lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge -f "The failing tests are not related."

fegin committed:
Update on "[SPMD] Upstream iter_move_grads_and_optimizers"


This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit to lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Upstream iter_move_grads_and_optimizers"


This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit to lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
